design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,315 @@
+
+ ## Non  Goals
+*  This features is not directly aimed for live migrating the VM using RWO volumes. Hence, the API covers and refers only to the volume migration.
+*  This design doesn't cover migration between storage classes, but rather implements a basic APIs that can be used by overlying tooling
+*  This design doesn't include the creation of the destination PVCs, it expects the destination PVCs to exist.
+
+ ## Definition of Users
+This feature addresses for all KubeVirt users who want to move the PVCs assigned to a VM without workload interruptions.
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated, and the volumes need to be moved to another storage class
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruption in order to give it more storage
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec VolumeMigrationSpec `json:"spec"`
+	Status *VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items []VolumeMigration `json:"items"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+	DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+	RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+	DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+	ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty" valid:"required"`
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type RejectedMigratedVolume struct {
+	SourcePvc string `json:"sourcePvc,omitempty" valid:"required"`
+	Reason    string `json:"reason,omitempty" valid:"required"`
+}
+
+type RejectedVolumeMigration struct {
+	RejectedVolumes []RejectedMigratedVolume `json:"rejectedVolumes,omitempty" valid:"required"`
+	AssociatedVolumes []string `json:"associatedVolumes,omitempty"`
+	Reason            string   `json:"reason,omitempty" valid:"required"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRejected   = "Rejected"
```

Comment:
Probably we do not need this phase anymore since we decided to fail the migration if there are rejected volumes.

Reply:
Yes, you are right I can simplify this a little bit a put a condition

Reply:
Done